includes/class-payment-information.php

```diff
@@ -32,31 +34,31 @@ class Payment_Information {
 	/**
 	 * Indicates whether the payment is merchant-initiated (true) or customer-initiated (false).
 	 *
-	 * @var bool
+	 * @var Payment_Initiated_By
 	 */
-	private $off_session;
+	private $payment_initiated_by;
 
 	/**
 	 * Payment information constructor.
 	 *
-	 * @param string            $payment_method The ID of the payment method used for this payment.
-	 * @param \WC_Payment_Token $token The payment token used for this payment.
-	 * @param bool              $off_session Indicates whether the payment is merchant-initiated (true) or customer-initiated (false).
+	 * @param string               $payment_method The ID of the payment method used for this payment.
+	 * @param \WC_Payment_Token    $token The payment token used for this payment.
+	 * @param Payment_Initiated_By $payment_initiated_by Indicates whether the payment is merchant-initiated or customer-initiated.
 	 *
 	 * @throws \Exception - If no payment method is found in the provided request.
 	 */
 	public function __construct(
 		string $payment_method,
 		\WC_Payment_Token $token = null,
-		bool $off_session = false
+		Payment_Initiated_By $payment_initiated_by = null
```

Comment:
@vbelolapotkov and I realized during some troubleshooting that default argument values in PHP [can't be the results of function calls](https://stackoverflow.com/a/10986437), so the default here is `null` and the default value is set in the constructor function body with `$payment_initiated_by ?? Payment_Initiated_By::CUSTOMER()`.
Just for the record; we still get type safety this way; you can only pass a `Payment_Initiated_By` or `null` to the constructor.

We could probably do a bit better here, e.g. by creating constants in the enum class, e.g.:

```php
class Payment_Initiated_By extends Enum {
	const MERCHANT = 'initiated_by_merchant';
	const CUSTOMER = 'initiated_by_customer';

	const Customer = new Payment_Initiated_By( self::CUSTOMER );
	const Merchant = new Payment_Initiated_By( self::MERCHANT );
};
```

and we could then (probably, I haven't tested this) initialize the constructor argument with `Payment_Initiated_By $payment_initiated_by = Payment_Initiated_By::Customer`.

Thoughts?

Reply:
I'd prefer constants approach (with `const Customer = new Payment_Initiated_By( self::CUSTOMER );`) rather than having `??` everywhere where we need fallback to default value. If it works :)

Reply:
Sadly, it seems my previously proposed solution does not work, and I can't find any variations on it that work either. So I think we're stuck with this version, unless we want to do something really elaborate 😕.

Reply:
Can we do something simple here like this?
```
class Payment_Initiated_By {
	const DEFAULT = new Payment_Initiated_By( 'initiated_by_customer' );
	
	private $value;

	public function __construct( $value ) {
		$this->validate( $value );
		$this->value = $value;
	}

	public function get_value() {
		return $this->value;
	}

	private function validate ( $value ) {
		if ( ! in_array( $value, ['initiated_by_merchant', 'initiated_by_customer'] ) ) {
			throw new Exception( 'Boom!' );
		}
	}
}
```

Reply:
This is a slightly more elaborate version of what I suggested [here](https://github.com/Automattic/woocommerce-payments/pull/898#discussion_r488448087), and could definitely work. All we'd really need to do is implement an `equals` function for this to be functionally equivalent to an `Enum`.

We would be losing the generic functionality of the `Enum` class though, and I think I'd rather keep the enums at the cost of a weird way to initialize variables, instead of re-implementing the same functionality across multiple different "enum" classes such as this one.

What do you think @dwainm @allendav @DanReyLop, since you've been part of this conversation as well?

Reply:
I don't think we need to check the value all over the place. This is a common pattern when null is a potential argument value. It can simply be checked at the last point before we need to use it.

I like that we're using the enum class as the approach works across multiple types. We can implement something simpler, but I don't see how we can improve on the API provided by the enum package.

My thinking would be to not reinvent the wheel and use the enum package as is, but I'm open to not using it as well.  The nice thing about the package is that we use most of the functionality.